src/ucs/memory/numa.c

```diff
@@ -72,3 +93,226 @@ int ucs_numa_node_of_cpu(int cpu)
+ * in @ref numa_distance_hash.
+ *
+ * @param [in]  distances NUMA distances list string
+ * @param [in]  source    source NUMA node
+ * @param [in]  dest      destination NUMA node
+ *
+ * @return Distance from @param source to @param dest
+ */
+static ucs_numa_distance_t ucs_numa_node_put_distances(char *distances,
+                                                       ucs_numa_node_t source,
+                                                       ucs_numa_node_t dest)
+{
+    ucs_numa_distance_t distance_to_dest = UCS_NUMA_MIN_DISTANCE;
+    ucs_numa_node_t node                 = 0;
+    ucs_numa_distance_t distance;
+    ucs_kh_put_t kh_put_status;
+    khiter_t hash_it;
+    char *token;
+
+    token = strtok(distances, UCS_NUMA_NODE_DISTANCE_DELIMITER);
+    while (token) {
+        distance = strtoul(token, 0, 0);
+        if (distance < UCS_NUMA_MIN_DISTANCE) {
+            ucs_debug("Parsed NUMA distance from %d to %d is out of range",
+                      source, node);
+            distance = UCS_NUMA_MIN_DISTANCE;
+        }
+
+        if (node == dest) {
+            distance_to_dest = distance;
+        }
+
+        hash_it = kh_put(numa_distance, &numa_distance_hash,
+                         ucs_numa_distance_hash_key(source, node),
+                         &kh_put_status);
+        if ((kh_put_status != UCS_KH_PUT_FAILED) &&
+            (kh_put_status != UCS_KH_PUT_KEY_PRESENT)) {
+            kh_value(&numa_distance_hash, hash_it) = distance;
+        }
+
+        node++;
+        token = strtok(NULL, UCS_NUMA_NODE_DISTANCE_DELIMITER);
+    }
+
+    return distance_to_dest;
+}
+
+ucs_numa_distance_t
+ucs_numa_distance(ucs_numa_node_t source, ucs_numa_node_t dest)
+{
+    ucs_numa_distance_t distance = UCS_NUMA_MIN_DISTANCE;
+    size_t buf_len               = 0;
+    ssize_t len                  = 0;
+    char *line                   = NULL;
+    char node_distance_path[PATH_MAX];
+    FILE *distance_fp;
+    khiter_t hash_it;
+
+    ucs_assert(source < ucs_numa_num_configured_nodes());
+    ucs_assert(dest < ucs_numa_num_configured_nodes());
+
+    hash_it = kh_get(numa_distance, &numa_distance_hash,
+                     ucs_numa_distance_hash_key(source, dest));
+
+    if (ucs_likely(hash_it != kh_end(&numa_distance_hash))) {
+        distance = kh_value(&numa_distance_hash, hash_it);
+    } else {
+        ucs_snprintf_safe(node_distance_path, PATH_MAX,
+                          UCS_NUMA_NODE_DISTANCE_PATH, source);
+        distance_fp = fopen(node_distance_path, "r");
+        if (!distance_fp) {
+            ucs_debug("Failed to open node %hd distance file", source);
+            return distance;
+        }
+
+        len = getline(&line, &buf_len, distance_fp);
+        fclose(distance_fp);
+        if (len <= 0) {
+            ucs_debug("Failed to parse node %hd distance file", source);
+        } else {
+            distance = ucs_numa_node_put_distances(line, source, dest);
```

Comment:
maybe rename ucs_numa_node_put_distances to ucs_numa_node_parse_distances, and move file open and read inside this function? IMO it will be more self-contained
also, could use stream operations like fscanf to read the nodes instead of parsing them from the string